### PR TITLE
Refactor update_checks into a more generic lit_autoupdate

### PR DESCRIPTION
**Pull request by jonmeow**
I've refactored the script in order to make it work in more contexts, which is why the delta is lost. I've actually refactored a significant amount with the intent of making the logic easier to understand, because I was also adjusting bits of it.

Some key notes:

- Removes the multi-pass update that was dealing with unfixed line numbers in explorer (I think the current script should work in one pass)
  - Fixed explorer to handle multiple line numbers on the same line (turns out we can rely on local format for line numbers).
- Using execv instead of imports because making Python imports work in a setup like this feels like it's not worth it; only a nuisance.
- Adding __init__.py to satisfy mypy, which otherwise considers the lit_autoupdate.py scripts to be issues.
- Using py because I was thinking sh would be more platform-dependent. py should port better to Windows.
- Getting rid of [[ID#]] capture groups in the semantics-ir tests because with the autoupdate it's kind of moot (also, hard to autogenerate the pairs without relying on the %### value).

Note this does mean tests switch to more of a "make a change, see which tests change" setup. I don't know that that's a _bad_ thing though -- it's pretty much how tests are being written right now, which is why I went down this rabbit hole. It's a nuisance to make a change then _manually_ have to update a bunch of code.

My intent is to use this for to convert parse-tree tests to lit, but I wanted to do this with _existing_ tests first as a proof of concept and to make sure there's agreement.

### CODE REVIEW

File: bazel/testing/lit_autoupdate_base.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+
+def replace_all(s: str, replacements: List[Tuple[str, str]]) -> str:
+    """Runs multiple replacements on a string."""
+    for before, after in replacements:
+        s = s.replace(before, after)
+    return s
+
+
+def get_actual_output(
```
Comment (zygoloid) [inline bazel/testing/lit_autoupdate_base.py]: Is there a more descriptive name we could use here? It's not clear to me what "actual" means here, especially given that we're doing some post-processing on the output. Would something like `get_matchable_test_output` be reasonable?

File: bazel/testing/lit_autoupdate_base.py [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+        return False
+
+    # Interleave the new CHECK: lines with the tested content.
+    with open(test, "w") as f:
+        f.writelines(formatted_result_lines)
+        return True
+
+
+def update_checks(update_args: UpdateArgs, tests: Set[str]) -> None:
+    """Runs bazel to update CHECK: lines in lit tests."""
```
Comment (zygoloid) [inline bazel/testing/lit_autoupdate_base.py]:
```suggestion
    """Updates CHECK: lines in lit tests."""
```
I don't think this runs `bazel`.

File: bazel/testing/lit_autoupdate_base.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        results = list(exec.map(map_helper, tests))
+
+    # Each update call indicates progress with a dot without a newline, so put a
+    # newline to wrap.
+    print(f"\nUpdated {results.count(True)} lit test(s).")
+
+
+def main() -> None:
+    # Go to the repository root so that paths will match bazel's view.
+    os.chdir(Path(__file__).parent.parent.parent)
```
Comment (zygoloid) [inline bazel/testing/lit_autoupdate_base.py]: Presumably this means that the test name arguments are interpreted relative to the repository root, which seems undesirable. Should we parse the args and `realpath` or `abspath` the `parsed_args.tests` list first?

**Review by zygoloid — APPROVED** [review]

File: bazel/testing/lit_autoupdate_base.py (hunk 1 above)
Comment (jonmeow) [reply bazel/testing/lit_autoupdate_base.py]: Done

File: bazel/testing/lit_autoupdate_base.py (hunk 3 above)
Comment (jonmeow) [reply bazel/testing/lit_autoupdate_base.py]: Done, with some shifts in logic to use pathlib.Path more.

Comment (jonmeow) [reply bazel/testing/lit_autoupdate_base.py]: (I felt it'd be clearer if parse_args() encapsulated more)

**Pull request merged**